### PR TITLE
Support widget attachments in scheduled email

### PR DESCRIPTION
**Pull request by tomassatinsky**
Jira: TNT-510

New UI and code for selecting and scheduling widget exports.
All new changes behind feature flag enableInsightExportScheduling

---

Supported PR commands:

| Command                  | Description             |
| ------------------------ | ----------------------- |
| `ok to test`             | Re-run standard checks  |
| `extended test`          | BackstopJS tests        |
| `extended check sonar`   | SonarQube tests         |
| `extended check cypress` | Cypress E2E tests       |
| `extended check plugins` | Dashboard plugins tests |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/7334/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9189/ : FAILURE in 8m 50s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/7364/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9209/ : FAILURE in 8m 44s

File: examples/playground/src/App.tsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// (C) 2019 GoodData Corporation
+// (C) 2019-2022 GoodData Corporation
```
Comment (scavnickyj) [inline examples/playground/src/App.tsx:1]: Revert the file and commit with no-verify flag please

File: libs/sdk-ui-dashboard/src/presentation/localization/bundles/en-US.json [hunk 2]
```diff
@@ -235,11 +240,41 @@
...
+    "dialogs.schedule.email.attachment.select.dashboard.header": {
+        "value": "Dashboard (PDF)",
+        "comment": "Heading of checkbox that selects dashboard to be exported in PDF file format",
+        "limit": 0
+    },
+    "dialogs.schedule.email.attachment.select.widgets.header": {
+        "value": "Insights (CSV, XLSX)",
+        "comment": "Heading of section of checkboxes for selecting insights to be exported as CSV of XLSX file",
+        "limit": 0
+    },
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/localization/bundles/en-US.json:267]: Missing translation for report. Also check other translations regarding this please

File: libs/sdk-backend-spi/src/common/settings.ts [hunk 3]
```diff
@@ -104,6 +104,11 @@ export interface ISettings {
      */
     enableKPIDashboardScheduleRecipients?: boolean;
 
+    /**
+     * Indicates whether the  the new UI and workflow for scheduled emailing with widgets is enabled.
```
Comment (scavnickyj) [inline libs/sdk-backend-spi/src/common/settings.ts:108]: fix wording here please

File: libs/sdk-backend-bear/src/convertors/fromBackend/DashboardConverter/tests/scheduledEmail.test.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+describe("scheduled mail converter", () => {
+    it("converts scheduled email object", () => {
+        const scheduledMail: GdcScheduledMail.IWrappedScheduledMail = {
+            scheduledMail: {
+                content: {
+                    when: {
+                        recurrency: "recurency string",
+                        startDate: "2022-01-01",
+                        timeZone: "CET",
+                        endDate: "2023-01-01", //?
```
Comment (scavnickyj) [inline libs/sdk-backend-bear/src/convertors/fromBackend/DashboardConverter/tests/scheduledEmail.test.ts:15]: what are those commented out question mark here?

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/Attachments.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+const AttachmentItem: React.FC<{ format: string }> = ({ format, children }) => (
+    <div className="gd-dashboard-attachment">
+        <span className="gd-dashboard-attachment-format">{format}</span>
+        <span className="gd-dashboard-attachment-name">{children}</span>
+    </div>
+);
+
+export type IAttachmentsProps = IAttachmentsOwnProps & WrappedComponentProps;
+
+const RenderAttachments = (props: IAttachmentsProps) => {
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/Attachments.tsx:35]: I would rename this to something like `AttachmentComponent`

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/AttachmentsSelectionDropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    insightWidgets: IInsightWidget[];
+    widgetsSelected: { [widgetUri: string]: boolean };
+    onApply(dashboardSelected: boolean, widgetsSelected: IWidgetsSelection): void;
+    theme?: ITheme;
+}
+
+export type IAttachmentsSelectionDropdownProps = IAttachmentsSelectionDropdownOwnProps &
+    WrappedComponentProps;
+
+const RenderAttachmentsSelectionDropdown: React.FC<IAttachmentsSelectionDropdownProps> = (props) => {
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/AttachmentsSelectionDropdown.tsx:31]: I would also rename this to not include the word `render`

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/FormatOptionsDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                        )}
+                    />
+                </div>
+            </div>
+        );
+    };
+
+    const renderBodyContentSelectorOptions = () => {
+        const attributeMergedMessage = intl.formatMessage({ id: "dialogs.schedule.email.attribute.merged" });
+        const showFiltersMessage = intl.formatMessage({ id: "dialogs.schedule.email.show.filters" });
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/FormatOptionsDropdown.tsx:97]: These could be replaced by `FormattedMessage` for simplicity. Also one more above

File: libs/sdk-ui-dashboard/styles/scss/scheduled_mail.scss [hunk 8]
```diff
@@ -306,3 +382,199 @@ $vertical-space: 10px;
...
+        display: flex;
+        flex: 0 0 auto;
+        align-items: center;
+        padding-left: 10px;
+        padding-right: 10px;
+        overflow: hidden;
+        height: 35px;
+        white-space: nowrap;
+        color: var(--gd-palette-complementary-7, #94a1ad);
+        background: var(--gd-palette-complementary-2, #ebeff4);
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/styles/scss/scheduled_mail.scss:446]: I believe that the fallback colors here are already in some of our variables. Please used them if possible. You may also find that there is already some variable with the same complementary and fallback. Use for others styles as well if applicable

e.g. `$is-focused-background`, `$default-gd-color-state-blank` or `$gd-color-state-blank` may be used?

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts [hunk 9]
```diff
@@ -123,20 +157,23 @@ export const useScheduledEmail = (props: UseScheduledEmailProps): UseScheduledEm
...
     const handleCreateScheduledEmail = useCallback(
         (scheduledEmail: IScheduledMailDefinition, customFilters?: FilterContextItem[]) => {
             // If dashboard filters are not changed, do not save them to scheduled email filter context.
             // Like this, future filter changes stored in the original dashboard filter context
             // are correctly propagated to the scheduled emails with the original filter context.
-            const filtersToStore = isEqual(originalFilters, filters) ? undefined : filters;
+            const filtersToStore = hasDefaultFilters ? undefined : filters;
             scheduledEmailCreator.create(scheduledEmail, customFilters ?? filtersToStore);
         },
         [filters, originalFilters],
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/useScheduledEmail.ts:179]: I believe the dependency array should also be updated now

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/ScheduledMailDialogRenderer.tsx [hunk 10]
```diff
@@ -442,6 +540,46 @@ export class ScheduledMailDialogRendererUI extends React.PureComponent<
         });
     };
 
+    // unit test
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/ScheduledMailDialogRenderer.tsx:543]: Is this a TODO note? I have not found the mentioned unit test anywhere..

File: libs/sdk-backend-bear/src/convertors/fromBackend/DashboardConverter/tests/scheduledEmail.test.ts (hunk 4 above)
Comment (tomassatinsky) [reply libs/sdk-backend-bear/src/convertors/fromBackend/DashboardConverter/tests/scheduledEmail.test.ts:15]: I forgot it here. It was just marking optional fields

File: libs/sdk-ui-dashboard/src/presentation/localization/bundles/en-US.json (hunk 2 above)
Comment (tomassatinsky) [reply libs/sdk-ui-dashboard/src/presentation/localization/bundles/en-US.json:267]: good catch!

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/ScheduledMailDialogRenderer.tsx (hunk 10 above)
Comment (tomassatinsky) [reply libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/ScheduledMailDialogRenderer.tsx:543]: Sorry, forgotten here. I've added unit tests that check the generated scheduled mail object

File: libs/sdk-ui-dashboard/styles/scss/scheduled_mail.scss (hunk 8 above)
Comment (tomassatinsky) [reply libs/sdk-ui-dashboard/styles/scss/scheduled_mail.scss:446]: I would suggest to try to fix it in the theming and responsiveness story. Anyway we will update the styles once again when we review the UI with Tomas Muchka. I can explicitly mention that in the story.

**Comment by gdgate** [comment]
Github ref error: SHA of pull request head checked out from remote PR head ref is different from head commit SHA from Github webhook event.

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/Attachments.tsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                        <AttachmentItem format="PDF">
+                            <span className="shortened-name">{dashboardTitle}</span>
+                        </AttachmentItem>
+                    )}
+                    {selectedWidgetsTitles.length !== 0 && (
+                        <AttachmentItem format={configuration.format}>
+                            <span className="shortened-name">{selectedWidgetsTitles.join(", ")}</span>
+                            <span>
+                                {selectedWidgetsTitles.length > 1 && `(${selectedWidgetsTitles.length})`}
+                            </span>
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/Attachments.tsx:69]: Is there some reason why the span can render empty text? If not, I would move the <span> into the condition.

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/AttachmentsSelectionDropdown.tsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+                            />
+                            <span title={dashboardTitle} className="input-label-text">
+                                {dashboardTitle}
+                            </span>
+                        </label>
+                    </div>
+                    <Typography tagName="h3">
+                        <FormattedMessage id="dialogs.schedule.email.attachment.select.widgets.header" />
+                    </Typography>
+                    {insightWidgets &&
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/Attachments/AttachmentsSelectionDropdown.tsx:100]: The condition here is not necessary as the array has empty fallback.

File: libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/test/ScheduledMailDialogRenderer.test.tsx [hunk 13]
```diff
@@ -100,72 +94,66 @@ describe("ScheduledMailDialogRenderer", () => {
         expect(onCancel).toHaveBeenCalledTimes(1);
     });
 
-    it("should trigger onSubmit on click Export", () => {
+    it("should generated scheduled mail with default values", () => {
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/test/ScheduledMailDialogRenderer.test.tsx:97]: generated -> generate

also in the test below

**Comment by scavnickyj** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Review by scavnickyj — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2365

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2127/ : SUCCESS in 1m 16s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2411/ : SUCCESS in 23s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1231/ : SUCCESS in 13m 46s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9214/ : FAILURE in 9m 03s

**Review by scavnickyj — APPROVED** [review]

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2128/ : SUCCESS in 1m 01s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2412/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1232/ : SUCCESS in 13m 12s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9217/ : FAILURE in 20m 29s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/7369/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9220/ : SUCCESS in 17m 51s

**Review by scavnickyj — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2365

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2130/ : SUCCESS in 1m 14s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2415/ : SUCCESS in 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1234/ : SUCCESS in 17m 34s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/9221/ : SUCCESS in 19m 40s

**Pull request merged**